### ISSUE TITLE
Requesting local earth rotation angle and one additional built-in frame

### ISSUE THREAD

**Issue by mkbrewer** [astropy/astropy#10239]
Just as I find it frustrating that the Time package provides local sidereal time, but the coordinate package doesn't currently include a built-in frame providing topocentric position with respect to the true equator and equinox of date (issue #10230), I find it equally frustrating that the Time package doesn't provide the local earth rotation angle to complement the coordinate package's built-in CIRS frame. When working with an equatorial mounted telescope both a set of equatorial coordinates and its complementary time or angle are required in order to calculate the hour angle necessary to point the telescope. The Time package should be enhanced to provide the local earth rotation angle. 

In this regard, it would also be nice to provide a built-in frame to transform CIRS coordinates to observed equatorial coordinates to complement the AltAz built-in frame. This should provide both declination and hour angle (or right ascension if the local earth rotation angle is available) adjusted for refraction and polar motion. This would be very simple as the same function that is used by AltAz (atioq()) provides these also.

**Comment by mkbrewer** [comment]
Observed declination and hour angle would probably be best here. These are the direct analog to azimuth and altitude and are independent of the origin of right ascension in use. It is quite easy to derive the observed right ascension in either frame from these. Subtracting the hour angle from the local sidereal time yields the observed right ascension referenced to the true equinox of date, while subtracting the hour angle from the local earth rotation angle yields the observed right ascension referenced to the CIO.

**Comment by mhvk** [comment]
@mkbrewer - rephrasing your request just so that I'm sure I understand them correctly.

1. Is the request to have a method, say, `Time.earth_rotation_angle(longitude)` that uses SOFA's `era00` and adds the longitude? I guess, like `sidereal_time`, it would return the result as a `Longitude` with units of hour-angle. This would be easy to do (and after `sidereal_time`, which I implemented, there will hopefully be less discussion on whether this belongs on `EarthLocation` or on `Time`!). Should there be any arguments beyond `longitude` (which would default to the `location` stored on the time)?

2. And for the coordinate frame, you are thinking of an "HADec" exactly analogous to "AltAz", but for polar mount? As you note, all that would seem to take is replicating `cirs_to_altaz` but catching different outputs in
 https://github.com/astropy/astropy/blob/35085f6d4db8e67b6794012b84e598eb94119212/astropy/coordinates/builtin_frames/cirs_observed_transforms.py#L66
The most difficult part here may be the name...

A more general comment on (2) is that between the various requests, it may be that quite a number of those frames in Fig.1 of the sofa tools docs might be useful on their own...

**Comment by MaxNoe** [comment]
I also just stumbled on this for our simulated gamma-ray events, which don't have a proper observation time and it was wished that the coordinates be not only expressed in AltAz but also in the local equatorial frame.

Having the earth_rotation_angle at `astropy.Time` seems indeed necessary, especially considering that this is "the way to go" now instead of using `sidereal_time`.

**Comment by mkbrewer** [comment]
@mhvk Sorry, I somehow missed the notification of your comment. There is one additional element to this, but it is very small. That's the TIO locator given by SOFA's `Sp00()`. The local earth rotation angle is then: 
```
Era00() + geodetic longitude + Sp00()
```
For reference see `Apco13()`, `Apco()`, `Aper13()` and `Aper()`.

**Comment by mhvk** [comment]
@mkbrewer - thanks! Just to be absolutely sure, there are no different "kinds" of earth rotation angle (like `kind='mean'` or `kind='apparent'` for sidereal time), nor is there a need for a possibly different model? I ask since if there are such differences, I'd rather mimic the argument structure of `sidereal_time(kind, longitude=None, model=None)` (though fine if there is a default for all arguments). 

I realize this all mostly comes from the fact that I have seen (and used) sidereal time a lot but have not encountered earth rotation angle before, and it is leaving me a bit confused that `apco` takes an earth rotation angle argument as well as an `sp`, i.e., with your equation, the earth rotation angle we would produce here for `longitude=0` would *not* be what one would put as input in `apco`. 

p.s. Checking wikipedia (what else...), https://en.wikipedia.org/wiki/Sidereal_time#Earth_Rotation_Angle gives a definition just like `era00`, without the `sp00` term, so if nothing else there will be others who will be confused as well... It also shows that my education well predates 2003...

**Comment by MaxNoe** [comment]
I'd have to check with  the explanatory supplements again, which are currently in my office and I am in the home office. 

But ERA00 should be Earth Rotation Angle at longitude 0, so I'm also a little confused that TIO comes into play here.

**Comment by mkbrewer** [comment]
Since the CIO is a non-rotating origin, it isn't subject to precession or nutation so there is no "mean" or "apparent" earth rotation angle. The models you refer to regarding sidereal time are just different nutation models so that isn't applicable. (see how much simpler things are when you use the ERA).  The local earth rotation angle that SOFA uses is ITRS, so it is referenced to the TIO. That's why the TIO locator is included. The SOFA "observed" coordinates are also ITRS, so this  is the local earth rotation angle can be used to derive the "observed" right ascension from the "observed" hour angle or vice versa, see `Atioq()`.

> I realize this all mostly comes from the fact that I have seen (and used) sidereal time a lot but have not encountered earth rotation angle before, and it is leaving me a bit confused that apco takes an earth rotation angle argument as well as an sp, i.e., with your equation, the earth rotation angle we would produce here for longitude=0 would not be what one would put as input in apco.

See `Apco13()` for the arguments that SOFA uses for `Apco()`. For longitude=0 or `'greenwich'` if you will, the ITRS local earth rotation angle is just: `Era00() + Sp00()`. See `astrom->along`. 

Of course, one could also define a TIRS local earth rotation angle for use with CIRS coordinates to derive the intermediate place hour angle also. This would need to add the offset described in equation (6.20) of USNO Circular 179. Or more appropriately: 
```
lamba_tirs = (sin(lamba_itrs) + y_p * tan(phi_itrs)) / (cos(lamba_itrs) - x_p * tan(phi_itrs))
```
which works all the way to the pole. But SOFA doesn't deal with this definition at all.

**Comment by mkbrewer** [comment]
By the way, in case you are interested, the way that SOFA handles polar motion is rather non-transparent. What they are doing is:
```
r_itrs = R_z(-lambda_itrs) W R_z(-ERA) r_cirs
```
Where `r_itrs` is the ITRS minus hour angle and declination, `W` is the wobble matrix and `r_cirs` is the CIRS right ascension and declination. This makes sense as the final rotation by geodetic longitude is applied after using the wobble matrix to transform to ITRS.

However, with their definition of local earth rotation angle, what they actually do is:
```
r_itrs = R_z(-lambda_itrs) W R_z(lambda_itrs + s') R_z(-ERAL) r_cirs
```
Where `s'` is the TIO locator. The expression ` R_z(-lambda_itrs) W R_z(lambda_itrs + s')` is evaluated in `Apco()` as "Polar motion, rotated onto the local meridian".

It took me awhile to figure that one out.

**Comment by mhvk** [comment]
@mkbrewer - I think I am starting to understand (having had to wrap my head around a few other problems hasn't helped). The confusion I had certainly stems from the SOFA implementation, where in `apco13`, an "earth rotation angle" is defined as
https://github.com/astropy/astropy/blob/399f7e11f6bef114ae0aa507cc6db8d8fead0534/cextern/erfa/apco13.c#L204-L205
and this angle is passed straight on to `apco`, where it is also called "earth rotation angle" and used in a call to `aper` to set `eral` or "local earth rotation angle" in
https://github.com/astropy/astropy/blob/399f7e11f6bef114ae0aa507cc6db8d8fead0534/cextern/erfa/aper.c#L96

So, yes, I can see that that takes figuring out! But that gets to the practical side, would it be useful to be able get the local earth rotation angle also in this form (which perhaps is nameless?). This especially since you mention that "TIRS" version which we could have to complement the "ITRS" one your request is really about. In any case, since our earth locations are all ITRS, that seems definitely the sensible default, so even if expand the options, I think `longitude` should always be the first argument, i.e., one would get a method `earth_rotation_angle(longitute=None, frame='itrs')`.

Finally, any preference for default units? `era00` itself is radians, but degrees may make more sense - at least the wikipedia article suggests that is used (rather than hours, a bit sadly, but probably wisely to avoid confusion).

**Comment by mkbrewer** [comment]
> So, yes, I can see that that takes figuring out! But that gets to the practical side, would it be useful to be able get the local earth rotation angle also in this form (which perhaps is nameless?).

I'm afraid that I don't understand the question here. What form are you talking about? 

> I think longitude should always be the first argument, i.e., one would get a method earth_rotation_angle(longitute=None, frame='itrs')

If your intention is to also support the TIRS local earth rotation angle, it will be necessary to have an option to pass the latitude also. Of course, this is the ITRS geocentric latitude rather than the geodetic latitude which adds further complication. The reason that you can use the geodetic longitude is that in the WGS84 model the geodetic longitude is the same as the geocentric longitude. This is not the case for latitude however. 

> Finally, any preference for default units?

This should definitely be the same as sidereal time (u.hourangle).

**Comment by mkbrewer** [comment]
One option is to only support TIRS local earth rotation angle if an `EarthLocation` is present, which I suspect would be true 99% of the time anyway. The SOFA function `Pvtob()` returns the TIRS local earth rotation angle directly, albeit as a Cartesian vector giving the observer's postion wrt the CIO. The longitudinal component of that vector is, in fact, the TIRS local earth rotation angle.

**Comment by mkbrewer** [comment]
I thought we were on the way here, but we seem to have stalled. In case that is over confusion between ITRS (observed) and TIRS (topocentic apparent or intermediate place), here is a function that returns the TIRS longitude and an example of one way to use it in the sidereal time function. It will work exactly the same way with the local earth rotation angle. I tried to do this in Python rather than my usual C. Please forgive any errors as I am not an expert at Python.
[tirs_longitude.py.txt](https://github.com/astropy/astropy/files/5139052/tirs_longitude.py.txt)

**Comment by mkbrewer** [comment]
This is just an update on my comment above on how SOFA handles polar motion. In my continuing investigation of this, I took this one step further. One can compute the observed ITRS right ascension and declination directly from the CIRS right ascension and declination by doing:
```
r_itrs = R_z(-ERAL) r_obs
```
Where `r_itrs` is the "observed" (ra, dec) and `r_obs` is the local ITRS observed (-ha, dec) as defined above. Due to the fact that, to first order, the wobble matrix is invariant under permutation, this reduces to:
```
r_itrs = R_z(-ERA) R_x(-yp) R_y(-xp) R_z(ERA) r_cirs
```
 In an analogy to "Polar motion, rotated onto the local meridian",  the expression `R_z(-ERA) R_x(-yp) R_y(-xp) R_z(ERA) ` can be defined at "Polar motion, rotated onto the CIO", by defining:
```
xpcio =  xp * cos(ERA) + yp * sin(ERA)
ypcio = -xp * sin(ERA) + yp * cos(ERA)
```
Then:
```
r_itrs = R_x(-xpcio) R_y(-ypcio) r_cirs
```
Note that, as expected, the TIO locator `s'` is nowhere to be found here. I tested this and it works perfectly. I do believe that it's rather elegant.

**Comment by maxnoe** [comment]
Even simpler would be to implement it as conversion from AltAz using:

https://pyerfa.readthedocs.io/en/latest/api/erfa.ae2hd.html#erfa.ae2hd

right?

**Comment by mkbrewer** [comment]
It's actually even simpler than that. The function that is use to derive AltAz also returns observed Hour Angle and Declination, so this frame can be realized in exactly the same way as AltAz. 

https://pyerfa.readthedocs.io/en/latest/api/erfa.atioq.html#erfa.atioq

**Comment by maxnoe** [comment]
Awesome. Do zou want to open the PR?

**Comment by mkbrewer** [comment]
I'm afraid I'm not enough of an astropy expert to do that. It's not so much the addition of a `cirs_to_hadec()` function to `cirs_observed_transforms` that's the problem. That's trivial. It's all the infrastructure that needs to be built up around a new frame that I don't really understand. There is also this issue of creating the required tests which I also no nothing about. Also, if I were to do this, I would want to address the serious error in AltAz (and by extension HADec) in regards to solar system bodies that arises from dropping the observer's location during the transform from topocentric GCRS coordinates to geocentric CIRS coordinates (thus dropping the aberration associated with diurnal parallax) and then tacking the observer's location back on during the conversion to observed coordinates. See issue #10356 for more on that.

**Comment by mkbrewer** [comment]
@mhvk, @StuartLittlefair  I'd really like to move ahead with this soon. I don't think there is any real reason to worry about the TIRS local earth rotation angle or sidereal time here. My investigation of polar motion showed that these result in incorrect observed coordinates due to the neglect of the angle `CIP->observer->ITRS pole` anyway. Thus, it is best to just ignore polar motion completely if the TIRS hour angle desired. So just returning `eral = era + longitude + TIO locator` should be fine.  While doing this `sidereal_time()` should also be modified to add the TIO locator for all models other than `IAU1982` and `IAU1994` as the TIO is the origin for these when local time is desired. In order to make it possible for the user to retrieve `era` or `gast`, the TIO locator could be left out if the longitude argument is `greenwich` or perhaps better `None`. 

Regarding the HA/Dec frame, I think the best thing to do is to rename `cirs_to_altaz()` `cirs_to_observed()`, give it an additional argument `'A'` or `'H'` and then create new functions `cirs_to_altaz()` and `cirs_to_hadec()` that call `cirs_to_observed()` with the appropriate argument. This should also apply to the reverse transforms.

**Comment by mhvk** [comment]
@mkbrewer - I think the hold-back here is simply that if we are deviating from what ERFA returns, we have to be super clear about why we are doing it (and why we are not trying to get any changes upstream...). This even more so if the results of `Time.sidereal_time` are going to change. Obviously, this has to start with wrapping one's head around the detailed comments you wrote above...

p.s. The HADec frame seems straightforward and your implementation suggestion sensible. Just a matter of someone finding time to actually do it.

**Comment by mkbrewer** [comment]
I don't see how we are deviating from anything that ERFA returns. The local earth rotation angle is exactly 
what `erfa.aper()` calculates. ERFA has no function to return the local sidereal time, but it's rather obvious what it should be. `GAST = ERA + EO`, where `ERA` is the angle between the CIO and the TIO and `EO` is the angle between the equinox and the CIO , so `GAST` is the angle between the equinox and the TIO. Since `s' = erfa.sp00()` is the angle between the TIO and the origin of longitude (the Greenwich meridian), `LST = GAST + longitude + s'`.

**Comment by mkbrewer** [comment]
Another way of looking at this is that `LST = ERAL + EO`.

**Comment by mhvk** [comment]
My worry was particularly about having to adjust `Time.sidereal_time`. Right now we simply call `gmst06` and `gst06a`, which state that they are "Greenwich mean/apparent sidereal time", but above you suggested that that should be adjusted. 

And I worry it may be similarly confusing is `earth_rotation_angle` does not just return the output of `era00`. 

Note that none of this is to suggest not to do it, just that it needs a super clear docstring, and, if I were to write it, I need to actually sit down and make drawings, etc., as right now I have only a vague sense.

**Comment by mkbrewer** [comment]
Sorry, I got the sign of the equation of the origins wrong there. I have now corrected it.

**Comment by mkbrewer** [comment]
Calling those two functions is fine. They give you GAST and GMST, which are the angles between the true and mean equinox of date and the TIO. It's just when you go to add a longitude you need to add the offset of the TIO from the Greenwich meridian. This is not true for `IAU1982` and `IAU1994` as the TIO isn't part of those models.

**Comment by mkbrewer** [comment]
This is why I said it would be best to use `longitude=None` for returning ERA, GAST or GMST. The local apparent and mean sidereal times at `greenwich` are actually `GAST + 0 + s'` and `GMST + 0 + s'`.

**Comment by mkbrewer** [comment]
I understand that this is a bit confusing. As a matter of tradition, the IAU still uses the terms Greenwich Apparent and Mean Sidereal Time even though they are no longer referenced to the Greenwich meridian.

**Comment by mhvk** [comment]
It is that last point that indeed is very confusing!! But with your explanations I at least understand a bit better why.... It would have been nice if this were pointed out in the comments of the ERFA functions.

**Comment by mkbrewer** [comment]
First line in `erfa.apco()`:
```
/* Longitude with adjustment for TIO locator s'. */
   astrom->along = elong + sp;
```

**Comment by mkbrewer** [comment]
If course, if the intended purpose of `sidereal_time('apparent', 'greenwich')` is to return GAST, then it should stay that way.

**Comment by mhvk** [comment]
No, definitely not. It is just that I took the name "Greenwich Apparent Sidereal Time" literally - it never occurred to me that "Greenwich" would not refer simply to the meridian... Now that you explained it is otherwise, I think your suggested solution of having `None` return the output of `gst06` directly (without offset) makes a lot of sense.

**Comment by mkbrewer** [comment]
@mhvk , @StuartLittlefair In version 1.36, Skyfield has added an `apparent.hadec()` method that can be directly compared to the Ha/Dec output of `erfa.atioq()`. This comparison is good to less than 1 mas:

![ha_dec](https://user-images.githubusercontent.com/10635444/106160495-e3fb2900-6153-11eb-8591-72348c262941.png)

and so can be used to test Astropy's `hadec` built-in frame (if it ever gets implemented).

**Comment by mhvk** [comment]
Nice! (And, yes, we should have the `HADec` built-in frame... It is always a matter of someone getting around to it...)

**Comment by mkbrewer** [comment]
@mhvk The new polar motion algorithm introduced in pyerfa 1.7.2 will affect the calculation of the local earth rotation angle and local sidereal time here. The change is minuscule as it is second order in the polar motion quantities.

**Comment by mkbrewer** [comment]
I can implement the local earth rotation angle also, but I'd need some guidance on how to retrieve TT and UT1 from the Time object in order to do so.

**Comment by maxnoe** [comment]
https://github.com/astropy/astropy/blob/339bb167086471ba6ce8d2c1831c3153fcd1dd03/astropy/coordinates/builtin_frames/utils.py#L96

**Comment by mhvk** [comment]
Fantastic! In principle, as simple as `time.tt` and `time.ut1`, and the implementation of `sidereal_time` should give some ideas, but perhaps best to deal with that after #11676? I'll request my own review on that one - hopefully some time next week...

**Comment by mkbrewer** [comment]
I'm guessing that this would be something along the lines of:
```
jd1 = getattr(getattr(self, 'tt')._time, 'jd1')
jd2 = getattr(getattr(self, 'tt')._time, 'jd2_filled')
```
Is that correct?

**Comment by mkbrewer** [comment]
My plan is to do this as a separate PR from #11676 as that will be cleaner.
